src/dune_rules/pkg_rules.ml

```diff
@@ -569,25 +597,25 @@ module Action_expander = struct
       (* TODO opam has a preprocessing step that we should probably apply *)
       Action.Run (patch, [ "-p1"; "-i"; input ])
     | Substitute (input, output) ->
-      let input =
-        Expander.expand_pform_gen ~mode:Single expander input
-        |> Value.to_path ~dir
-      in
-      let output =
-        Expander.expand_pform_gen ~mode:Single expander output
-        |> Value.to_path ~dir |> Path.as_in_build_dir_exn
-      in
-      Memo.return @@ Substitute.action ~input ~output
+      let* input = Expander.expand_pform_gen ~mode:Single expander input in
+      let input = input |> Value.to_path ~dir in
+      let+ output = Expander.expand_pform_gen ~mode:Single expander output in
+      let output = output |> Value.to_path ~dir |> Path.as_in_build_dir_exn in
+      Substitute.action ~input ~output
     | Withenv (updates, action) ->
-      let+ action = expand action ~expander in
-      let _env, updates =
-        List.fold_left ~init:(expander.env, []) updates
-          ~f:(fun (env, updates) ({ Env_update.op = _; var; value } as update)
-             ->
-            let value =
+      let* action = expand action ~expander in
+      let+ _env, updates =
+        List.fold_left
```

Comment:
Memo.List.fold_left

Reply:
Ah, I was looking for exactly this! I overlooked that `Memo` includes `S` which has `List` when skimming through the mli.